zomes/chat/src/lib.rs

```diff
@@ -61,15 +74,43 @@ fn init(_: ()) -> ExternResult<InitCallbackResult> {
         access: ().into(),
         functions,
     })?;
+    let entries = &query(ChainQueryFilter::new().header_type(HeaderType::AgentValidationPkg))?;
+    if let Header::AgentValidationPkg(h) = entries[0].header() {
+        match &h.membrane_proof {
+            Some(mem_proof) => {
+                if validation::is_read_only_proof(&mem_proof) {
+                    return Ok(InitCallbackResult::Pass)
+                }
+                let mem_proof = match Element::try_from(mem_proof.clone()) {
+                    Ok(m) => m,
+                    Err(_e) => return  Err(ChatError::InitFailure.into())
+                };
+                let code = mem_proof.entry().to_app_option::<validation::JoiningCodePayload>()?.unwrap();
+                trace!("looking for {:?}", code.record_locator);
+                let path = Path::from(code.record_locator.clone());
+                if path.exists()? {
+                    return Ok(InitCallbackResult::Fail(format!("membrane proof for {} already used", code.record_locator)))
+                }
+                trace!("creating {:?}", code.record_locator);
+                path.ensure()?;
+            },
+            None => return Err(ChatError::InitFailure.into()),
+        }
+    } else {
+        return Err(ChatError::InitFailure.into());
+    }
 
     Ok(InitCallbackResult::Pass)
 }
 #[hdk_extern]
 fn genesis_self_check(data: GenesisSelfCheckData) -> ExternResult<ValidateCallbackResult> {
-    validation::joining_code(data.membrane_proof)
+    validation::joining_code(data.agent_key, data.membrane_proof, true)
 }
 #[hdk_extern]
 fn create_channel(channel_input: ChannelInput) -> ExternResult<ChannelData> {
+    if is_read_only_instance() {
```

Comment:
Just as good practice shouldn't we put this in the validation on committing and channel entry? So that other agents validate it as well?

But since this is a temporary solution I can go either way

Reply:
Yes, we probably should, but there it's a network get of the agent's membrane-proof, which I thought wasn't worth slowing us down at this point, since in the future we wouldn't do this at all anyway.